python/tk_multi_publish2/api/task.py

```diff
@@ -0,0 +1,174 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import uuid
+
+import sgtk
+from .plugins import Setting, PublishPluginInstance
+
+logger = sgtk.platform.get_logger(__name__)
+
+
+class PublishTask(object):
+    """
+    A task is a particular unit of work which can to be carried by the
+    publishing process. A task can be thought of as a 'plugin instance', or a
+    particular publishing plugin operating on a collected item.
+    """
+
+    @classmethod
+    def from_dict(cls, task_dict, item=None):
+
+        # create the plugin instance here...
+        plugin = PublishPluginInstance.from_dict(task_dict["plugin"])
+
+        new_task = PublishTask(plugin, item)
+        new_task._name = task_dict["name"]
+        new_task._description = task_dict["description"]
+        new_task._active = task_dict["active"]
+        new_task._visible = task_dict["visible"]
+        new_task._enabled = task_dict["enabled"]
+
+        for (k, setting) in task_dict["settings"].iteritems():
+            new_setting = Setting(
+                setting["name"],
+                setting["type"],
+                setting["default_value"],
+                setting["description"]
+            )
+            new_setting.value = setting["value"]
```

Comment:
Didn't you point out there other day that the value could be anything?

Maybe we could so something where we serialize builtin types like collections, numbers and strings as is in the document and pickle everything else?

Reply:
So the json would contain value or pickled_value depending on how it was serialized?

Reply:
Keep in mind that when serializing collections, you need to recurse into them to know if there are non easily serializable objects.

I just realized that stuff like sets would need to be pickled since turning them into an array loses their semantic.

Maybe we should just always pickled everything? In any case, let's use pickled_value in the json document, just in case in the future we need to store unpickled versions of certain values.

Reply:
These are the "settings" which have to be representable via the config anyway. So I think we're ok here. But your points are valid for the local/global properties on the item. Let me have another look to see what we should do here. I'm assuming this will be a major version release, so maybe we should just add a restriction as to what can be shoved in there.

Reply:
Sorry, I got the two confused, you're totally right. I meant the global/local properties.